src/gdft.c

```diff
@@ -267,168 +260,172 @@ gdMutexDeclare (gdFontCacheMutex);
 static gdCache_head_t *fontCache;
 static FT_Library library;
 
-#define Tcl_UniChar int
-#define TCL_UTF_MAX 3
-static int
-gdTcl_UtfToUniChar (const char *str, Tcl_UniChar * chPtr)
-/* str is the UTF8 next character pointer */
-/* chPtr is the int for the result */
-{
-	int byte;
-	char entity_name_buf[ENTITY_NAME_LENGTH_MAX+1];
-	char *p;
-	struct entities_s key, *res;
-
-	/* HTML4.0 entities in decimal form, e.g. &#197; */
-	/*           or in hexadecimal form, e.g. &#x6C34; */
-	byte = *((unsigned char *) str);
-	if (byte == '&') {
-		int i, n = 0;
-
-		byte = *((unsigned char *) (str + 1));
-		if (byte == '#') {
-			byte = *((unsigned char *) (str + 2));
-			if (byte == 'x' || byte == 'X') {
-				for (i = 3; i < 8; i++) {
-					byte = *((unsigned char *) (str + i));
-					if (byte >= 'A' && byte <= 'F')
-						byte = byte - 'A' + 10;
-					else if (byte >= 'a' && byte <= 'f')
-						byte = byte - 'a' + 10;
-					else if (byte >= '0' && byte <= '9')
-						byte = byte - '0';
-					else
-						break;
-					n = (n * 16) + byte;
-				}
-			} else {
-				for (i = 2; i < 8; i++) {
-					byte = *((unsigned char *) (str + i));
-					if (byte >= '0' && byte <= '9')
-						n = (n * 10) + (byte - '0');
-					else
-						break;
-				}
+static int gd_Entity_To_Unicode(const char *str, uint32_t *chPtr){
+	// str is the UTF8 next character pointer, or a pointer to a string of 4 different kinds of html entity notation
+	// chPtr is the int for the result
+	// return value is the offset from str to the char for the next search to start
+	// function falls through and returns 0 if no valid entity detected, leaving chPtr alone.
+	// HTML4.0/5.0 entities in hexadecimal form (e.g. &#x00021; and &#X00021;), in decimal form (e.g. &#33;),  or in string form (e.g. &excl;)
+	char entity_name_buf[ENTITY_NAME_LENGTH_MAX];
+	uint32_t n = 0;
+	uint32_t i = 1;
+	uint32_t b0 = *(uint8_t *)str;
```

Comment:
you immediately clobber this assignment below, so delete this

Reply:
n needs to be 0.

Reply:
i'm not referring to `n`, i'm referring to `b0`.

that said, i was thinking `b0` was a pointer you were using to walk `str` rather than a char you've pulled out, so nm me here.